File: apps/ruby-sdk/lib/firecrawl/models/monitor.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # A scrape, crawl, or search target stored on a monitor.
    #
    # Targets are forwarded to the API verbatim, so callers may also pass
    # plain Hashes with camelCase keys. This class is a convenience for
    # building well-formed target payloads (and is what {#to_h} emits).
    #
    # Search targets (+type: "search"+) use the +queries+, +search_window+,
    # +include_domains+, +exclude_domains+, and +max_results+ fields.
    class MonitorTarget
      attr_reader :id, :type, :urls, :url, :scrape_options, :crawl_options,
                  :queries, :search_window, :include_domains,
                  :exclude_domains, :max_results

      def initialize(data)
        @id = data["id"]
        @type = data["type"]
        @urls = data["urls"]
        @url = data["url"]
        @scrape_options = data["scrapeOptions"]
        @crawl_options = data["crawlOptions"]
        # search target fields
        @queries = data["queries"]
        @search_window = data["searchWindow"]
        @include_domains = data["includeDomains"]
        @exclude_domains = data["excludeDomains"]
        @max_results = data["maxResults"]
      end

      def to_h
        {
          "id" => @id,
          "type" => @type,
          "urls" => @urls,
          "url" => @url,
          "scrapeOptions" => @scrape_options,
          "crawlOptions" => @crawl_options,
          "queries" => @queries,
          "searchWindow" => @search_window,
          "includeDomains" => @include_domains,
          "excludeDomains" => @exclude_domains,
          "maxResults" => @max_results,
        }.compact
      end
    end

    # A per-target result on a monitor check.
    #
    # Search targets (+type: "search"+) populate the +search_completed+,
    # +result_count+, +matches+, +summary+, +judge_degraded+,
    # +degraded_reason+, +search_credits+, +judge_credits+, and
    # +results_judged+ fields.
    class MonitorTargetResult
      attr_reader :target_id, :type, :expected_jobs, :crawl_id,
                  :search_completed, :result_count, :matches, :summary,
                  :judge_degraded, :degraded_reason, :search_credits,
                  :judge_credits, :results_judged

      def initialize(data)
        @target_id = data["targetId"]
        @type = data["type"]
        @expected_jobs = data["expectedJobs"]
        @crawl_id = data["crawlId"]
        # search target result fields
        @search_completed = data["searchCompleted"]
        @result_count = data["resultCount"]
        @matches = data["matches"]
        @summary = data["summary"]
        @judge_degraded = data["judgeDegraded"]
        @degraded_reason = data["degradedReason"]
        @search_credits = data["searchCredits"]
        @judge_credits = data["judgeCredits"]
        @results_judged = data["resultsJudged"]
      end
    end

    class Monitor
      attr_reader :id, :name, :status, :schedule, :next_run_at, :last_run_at,
                  :current_check_id, :targets, :webhook, :notification,
                  :retention_days, :estimated_credits_per_month,
                  :last_check_summary, :goal, :judge_enabled,
                  :created_at, :updated_at

      def initialize(data)
        @id = data["id"]
        @name = data["name"]
        @status = data["status"]
        @schedule = data["schedule"]
        @next_run_at = data["nextRunAt"]
        @last_run_at = data["lastRunAt"]
        @current_check_id = data["currentCheckId"]
        @targets = data["targets"] || []
        @webhook = data["webhook"]
        @notification = data["notification"]
        @retention_days = data["retentionDays"]
        @estimated_credits_per_month = data["estimatedCreditsPerMonth"]
        @last_check_summary = data["lastCheckSummary"]
        @goal = data["goal"]
        @judge_enabled = data["judgeEnabled"] || false
        @created_at = data["createdAt"]
        @updated_at = data["updatedAt"]
      end
    end

    class MonitorCheck
      attr_reader :id, :monitor_id, :status, :trigger, :scheduled_for,
                  :started_at, :finished_at, :estimated_credits,
                  :reserved_credits, :actual_credits, :billing_status,
                  :summary, :target_results, :notification_status, :error,
                  :created_at, :updated_at

      def initialize(data)
        @id = data["id"]
        @monitor_id = data["monitorId"]
        @status = data["status"]
        @trigger = data["trigger"]
        @scheduled_for = data["scheduledFor"]
        @started_at = data["startedAt"]
        @finished_at = data["finishedAt"]
        @estimated_credits = data["estimatedCredits"]
        @reserved_credits = data["reservedCredits"]
        @actual_credits = data["actualCredits"]
        @billing_status = data["billingStatus"]
        @summary = data["summary"] || {}
        @target_results = data["targetResults"]
        @notification_status = data["notificationStatus"]
        @error = data["error"]
        @created_at = data["createdAt"]
        @updated_at = data["updatedAt"]
      end
    end

    # A monitor check with paginated page results inlined.
    #
    # Each entry in {#pages} is a {Hash} with the standard monitor page
    # fields (id, targetId, url, status, previousScrapeId,
    # currentScrapeId, statusCode, error, metadata, createdAt) plus:
    #
    # * +"diff"+  – when the page changed. A hash with +"text"+ (markdown
    #   unified diff) and/or +"json"+ (parseDiff AST for markdown
    #   monitors, or a per-field +{ "previous", "current" }+ map for
    #   JSON-extraction monitors).
    # * +"snapshot"+ – present on JSON / mixed-mode monitors. A hash
    #   with a +"json"+ key holding the current JSON extraction at this
    #   run.
    # * +"judgment"+ – present when the monitor has a +goal+ set and
    #   judging is enabled. A hash with +"meaningful"+ (Boolean),
    #   +"confidence"+ (+"high"+ / +"medium"+ / +"low"+), +"reason"+
    #   (String), and +"fields"+ (Array of String) describing which
    #   fields the judge weighed.
    class MonitorCheckDetail < MonitorCheck
      attr_accessor :pages, :next_url

      def initialize(data)
        super
        @pages = data["pages"] || []
        @next_url = data["next"]
      end
    end
  end
end

File: apps/rust-sdk/src/monitor.rs
//! Monitor endpoint for Firecrawl API v2.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::client::Client;
use crate::FirecrawlError;

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct MonitorSchedule {
    pub cron: String,
    pub timezone: Option<String>,
}

#[serde_with::skip_serializing_none]
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CreateMonitorRequest {
    pub name: String,
    pub schedule: MonitorSchedule,
    pub targets: Vec<Value>,
    pub webhook: Option<Value>,
    pub notification: Option<Value>,
    pub retention_days: Option<u32>,
}

#[serde_with::skip_serializing_none]
#[derive(Deserialize, Serialize, Debug, Default, Clone)]
#[serde(rename_all = "camelCase")]
pub struct UpdateMonitorRequest {
    pub name: Option<String>,
    pub status: Option<String>,
    pub schedule: Option<MonitorSchedule>,
    pub targets: Option<Vec<Value>>,
    pub webhook: Option<Value>,
    pub notification: Option<Value>,
    pub retention_days: Option<u32>,
}

#[derive(Deserialize, Serialize, Debug, Default, Clone)]
#[serde(rename_all = "camelCase")]
pub struct MonitorSummary {
    pub total_pages: u32,
    pub same: u32,
    pub changed: u32,
    pub new: u32,
    pub removed: u32,
    pub error: u32,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Monitor {
    pub id: String,
    pub name: String,
    pub status: String,
    pub schedule: MonitorSchedule,
    pub next_run_at: Option<String>,
    pub last_run_at: Option<String>,
    pub current_check_id: Option<String>,
    pub targets: Vec<Value>,
    pub webhook: Option<Value>,
    pub notification: Option<Value>,
    pub retention_days: u32,
    pub estimated_credits_per_month: Option<u32>,
    pub last_check_summary: Option<MonitorSummary>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct MonitorCheck {
    pub id: String,
    pub monitor_id: String,
    pub status: String,
    pub trigger: String,
    pub scheduled_for: Option<String>,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub estimated_credits: Option<u32>,
    pub reserved_credits: Option<u32>,
    pub actual_credits: Option<u32>,
    pub billing_status: String,
    pub summary: MonitorSummary,
    pub target_results: Option<Value>,
    pub notification_status: Option<Value>,
    pub error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Per-field diff entry returned for monitors that requested JSON extraction.
#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct MonitorJsonFieldDiff {
    pub previous: Value,
    pub current: Value,
}

/// Diff payload returned alongside a monitor page when its scrape produced
/// a change. The shape depends on what the monitor's formats asked for:
///
/// - markdown-only monitors  → `text` is the unified diff and `json` is
///   the `parseDiff` AST (a `{ "files": [...] }` object).
/// - JSON-extraction monitors → `json` is the per-field
///   `{ previous, current }` map and `text` is absent.
/// - mixed (JSON + git-diff) monitors → both `text` (markdown sidecar)
///   and `json` (field-level diff) are present.
///
/// `json` is kept as a raw [`serde_json::Value`] so callers can decode it
/// into either shape (`HashMap<String, MonitorJsonFieldDiff>` for the
/// field-diff case, or a `{ files: [...] }` struct for the parseDiff AST).
#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct MonitorPageDiff {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub json: Option<Value>,
}

/// Snapshot of the current JSON extraction at this run. Present on JSON
/// and mixed-mode monitors; absent for markdown-only.
#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct MonitorPageSnapshot {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub json: Option<Value>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct MonitorCheckPage {
    pub id: String,
    pub target_id: String,
    pub url: String,
    pub status: String,
    pub previous_scrape_id: Option<String>,
    pub current_scrape_id: Option<String>,
    pub status_code: Option<u16>,
    pub error: Option<String>,
    pub metadata: Option<Value>,
    pub diff: Option<MonitorPageDiff>,
    pub snapshot: Option<MonitorPageSnapshot>,
    pub created_at: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct MonitorCheckDetail {
    #[serde(flatten)]
    pub check: MonitorCheck,
    pub pages: Vec<MonitorCheckPage>,
    pub next: Option<String>,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
struct DataResponse<T> {
    data: T,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
struct SuccessResponse {
    success: bool,
}

fn query(limit: Option<u32>, offset: Option<u32>, status: Option<&str>) -> String {
    let mut params = Vec::new();
    if let Some(limit) = limit {
        params.push(format!("limit={}", limit));
    }
    if let Some(offset) = offset {
        params.push(format!("offset={}", offset));
    }
    if let Some(status) = status {
        params.push(format!("status={}", status));
    }
    if params.is_empty() {
        String::new()
    } else {
        format!("?{}", params.join("&"))
    }
}

fn check_page_query(limit: Option<u32>, skip: Option<u32>, status: Option<&str>) -> String {
    let mut params = Vec::new();
    if let Some(limit) = limit {
        params.push(format!("limit={}", limit));
    }
    if let Some(skip) = skip {
        params.push(format!("skip={}", skip));
    }
    if let Some(status) = status {
        params.push(format!("status={}", status));
    }
    if params.is_empty() {
        String::new()
    } else {
        format!("?{}", params.join("&"))
    }
}

impl Client {
    pub async fn create_monitor(
        &self,
        request: CreateMonitorRequest,
    ) -> Result<Monitor, FirecrawlError> {
        let response = self
            .client
            .post(self.url("/monitor"))
            .headers(self.prepare_headers(None))
            .json(&request)
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Creating monitor".to_string(), e))?;

        let response: DataResponse<Monitor> =
            self.handle_response(response, "create monitor").await?;
        Ok(response.data)
    }

    pub async fn list_monitors(
        &self,
        limit: Option<u32>,
        offset: Option<u32>,
    ) -> Result<Vec<Monitor>, FirecrawlError> {
        let response = self
            .client
            .get(self.url(&format!("/monitor{}", query(limit, offset, None))))
            .headers(self.prepare_headers(None))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Listing monitors".to_string(), e))?;

        let response: DataResponse<Vec<Monitor>> =
            self.handle_response(response, "list monitors").await?;
        Ok(response.data)
    }

    pub async fn get_monitor(
        &self,
        monitor_id: impl AsRef<str>,
    ) -> Result<Monitor, FirecrawlError> {
        let response = self
            .client
            .get(self.url(&format!("/monitor/{}", monitor_id.as_ref())))
            .headers(self.prepare_headers(None))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Getting monitor".to_string(), e))?;

        let response: DataResponse<Monitor> = self.handle_response(response, "get monitor").await?;
        Ok(response.data)
    }

    pub async fn update_monitor(
        &self,
        monitor_id: impl AsRef<str>,
        request: UpdateMonitorRequest,
    ) -> Result<Monitor, FirecrawlError> {
        let response = self
            .client
            .patch(self.url(&format!("/monitor/{}", monitor_id.as_ref())))
            .headers(self.prepare_headers(None))
            .json(&request)
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Updating monitor".to_string(), e))?;

        let response: DataResponse<Monitor> =
            self.handle_response(response, "update monitor").await?;
        Ok(response.data)
    }

    pub async fn delete_monitor(
        &self,
        monitor_id: impl AsRef<str>,
    ) -> Result<bool, FirecrawlError> {
        let response = self
            .client
            .delete(self.url(&format!("/monitor/{}", monitor_id.as_ref())))
            .headers(self.prepare_headers(None))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Deleting monitor".to_string(), e))?;

        let response: SuccessResponse = self.handle_response(response, "delete monitor").await?;
        Ok(response.success)
    }

    pub async fn run_monitor(
        &self,
        monitor_id: impl AsRef<str>,
    ) -> Result<MonitorCheck, FirecrawlError> {
        let response = self
            .client
            .post(self.url(&format!("/monitor/{}/run", monitor_id.as_ref())))
            .headers(self.prepare_headers(None))
            .json(&serde_json::json!({}))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Running monitor".to_string(), e))?;

        let response: DataResponse<MonitorCheck> =
            self.handle_response(response, "run monitor").await?;
        Ok(response.data)
    }

    pub async fn list_monitor_checks(
        &self,
        monitor_id: impl AsRef<str>,
        limit: Option<u32>,
        offset: Option<u32>,
    ) -> Result<Vec<MonitorCheck>, FirecrawlError> {
        let path = format!(
            "/monitor/{}/checks{}",
            monitor_id.as_ref(),
            query(limit, offset, None)
        );
        let response = self
            .client
            .get(self.url(&path))
            .headers(self.prepare_headers(None))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Listing monitor checks".to_string(), e))?;

        let response: DataResponse<Vec<MonitorCheck>> = self
            .handle_response(response, "list monitor checks")
            .await?;
        Ok(response.data)
    }

    pub async fn get_monitor_check(
        &self,
        monitor_id: impl AsRef<str>,
        check_id: impl AsRef<str>,
        limit: Option<u32>,
        skip: Option<u32>,
        status: Option<&str>,
    ) -> Result<MonitorCheckDetail, FirecrawlError> {
        let path = format!(
            "/monitor/{}/checks/{}{}",
            monitor_id.as_ref(),
            check_id.as_ref(),
            check_page_query(limit, skip, status)
        );
        let response = self
            .client
            .get(self.url(&path))
            .headers(self.prepare_headers(None))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Getting monitor check".to_string(), e))?;

        let response: DataResponse<MonitorCheckDetail> =
            self.handle_response(response, "get monitor check").await?;
        let mut check = response.data;

        while let Some(next) = check.next.clone() {
            let response = self
                .client
                .get(next)
                .headers(self.prepare_headers(None))
                .send()
                .await
                .map_err(|e| {
                    FirecrawlError::HttpError("Getting monitor check page".to_string(), e)
                })?;
            let response: DataResponse<MonitorCheckDetail> = self
                .handle_response(response, "get monitor check page")
                .await?;
            check.pages.extend(response.data.pages);
            check.next = response.data.next;
        }

        Ok(check)
    }

    pub async fn get_monitor_check_page(
        &self,
        monitor_id: impl AsRef<str>,
        check_id: impl AsRef<str>,
        limit: Option<u32>,
        skip: Option<u32>,
        status: Option<&str>,
    ) -> Result<MonitorCheckDetail, FirecrawlError> {
        let path = format!(
            "/monitor/{}/checks/{}{}",
            monitor_id.as_ref(),
            check_id.as_ref(),
            check_page_query(limit, skip, status)
        );
        let response = self
            .client
            .get(self.url(&path))
            .headers(self.prepare_headers(None))
            .send()
            .await
            .map_err(|e| FirecrawlError::HttpError("Getting monitor check".to_string(), e))?;

        let response: DataResponse<MonitorCheckDetail> =
            self.handle_response(response, "get monitor check").await?;
        Ok(response.data)
    }
}
